Networking/WebSockets/WebSocketProtocol.hh

```diff
@@ -332,28 +337,28 @@ namespace uWS {
             memcpy(dst + headerLength, src, length);
 
             if ( !isServer ) {
-#if 1
-                // overwrites up to 3 bytes outside of the given buffer!
-                //WebSocketProtocol<isServer>::unmaskInplace(dst + headerLength, dst + headerLength + length, mask);
-
-                // this is not optimal
-                std::byte* start = dst + headerLength;
-                std::byte* stop  = start + length;
-                int        i     = 0;
-                while ( start != stop ) { (*start++) ^= mask[i++ % 4]; }
-#else
-                uint32_t mask_i32 = *reinterpret_cast<uint32_t*>(mask);
+                std::byte* start_byte = dst + headerLength;
+                std::byte* end_byte   = start_byte + length;
+                size_t     offset     = 0;
 
-                auto*     start = reinterpret_cast<uint32_t*>(dst + headerLength);
-                uint32_t* end   = start + length / 4;
+                // Handle first x amount of bytes individually until we are aligned with the 4-byte alignment
+                for ( ; (reinterpret_cast<uintptr_t>(start_byte) & 3) != 0 && start_byte != end_byte; ++offset ) {
+                    *start_byte++ ^= mask[offset % 4];
+                }
 
+                // Rotate mask by the offset we reached in the first loop to become memory-aligned
+                std::rotate(mask.begin(), mask.begin() + offset, mask.end());
+
+                // Process majority of bytes in chunks of 4 (until end or < 4 bytes away from end)
+                auto*     start = reinterpret_cast<uint32_t*>(start_byte);
+                uint32_t* end   = start + ((end_byte - start_byte) / 4);
+                uint32_t  mask_i32{};
+                std::memcpy(&mask_i32, mask.data(), 4);
                 while ( start != end ) { *start++ ^= mask_i32; }
```

Comment:
It's so complicated. mask has been adjusted for endianness, etc, to reduce the time by factor of 4. How to get it fully tested?

Reply:
WebSocket has built-in "testing", which I discovered during the painful process of getting all this working. If the data received through WebSocket isn't perfect, it will give some warning. A couple examples were "Frame too small" or "Invalid checksum", and there were a few more. I think the main method I used for testing was "P2P Sync" test.